Package request: phantomjs

Hello,

It would be nice to have a headless browser: http://phantomjs.org/

> Looks like there is a qt build for termux (https://github.com/termux/x11-packages/blob/master/packages/qt5-base/build.sh) so maybe at some point someone can port a build script for phantomjs

Well, those patches are certainly useful, in getting a little further building this highly defective source.

After applying those patches, reminding the linker that ``-latomic`` does actually exist (contrary to what it would like you to think) in the .pri files, and having to use ``termux-fix-shebang`` to even be able to get the initial errors, I am left with this devastating malfunction:


```
ssl/qsslcertificate_openssl.cpp:63:30: error: member access into incomplete
      type 'X509' (aka 'x509_st')
        return qHashBits(x509->sha1_hash, SHA_DIGEST_LENGTH, seed);
                             ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:120:16: note: 
      forward declaration of 'x509_st'
typedef struct x509_st X509;
               ^
ssl/qsslcertificate_openssl.cpp:87:68: error: member access into incomplete
      type 'X509' (aka 'x509_st')
            QByteArray::number(qlonglong(q_ASN1_INTEGER_get(d->x509->...
                                                                   ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:120:16: note: 
      forward declaration of 'x509_st'
typedef struct x509_st X509;
               ^
ssl/qsslcertificate_openssl.cpp:96:45: error: member access into incomplete
      type 'X509' (aka 'x509_st')
        ASN1_INTEGER *serialNumber = d->x509->cert_info->serialNumber;
                                            ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:120:16: note: 
      forward declaration of 'x509_st'
typedef struct x509_st X509;
               ^
ssl/qsslcertificate_openssl.cpp:232:32: error: member access into incomplete
      type 'X509' (aka 'x509_st')
    X509_PUBKEY *xkey = d->x509->cert_info->key;
                               ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:120:16: note: 
      forward declaration of 'x509_st'
typedef struct x509_st X509;
               ^
ssl/qsslcertificate_openssl.cpp:236:29: error: member access into incomplete
      type 'EVP_PKEY' (aka 'evp_pkey_st')
    if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_RSA) {
                            ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:93:16: note: 
      forward declaration of 'evp_pkey_st'
typedef struct evp_pkey_st EVP_PKEY;
               ^
ssl/qsslcertificate_openssl.cpp:240:36: error: member access into incomplete
      type 'EVP_PKEY' (aka 'evp_pkey_st')
    } else if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_DSA) {
                                   ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:93:16: note: 
      forward declaration of 'evp_pkey_st'
typedef struct evp_pkey_st EVP_PKEY;
               ^
ssl/qsslcertificate_openssl.cpp:245:36: error: member access into incomplete
      type 'EVP_PKEY' (aka 'evp_pkey_st')
    } else if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_EC) {
                                   ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:93:16: note: 
      forward declaration of 'evp_pkey_st'
typedef struct evp_pkey_st EVP_PKEY;
               ^
ssl/qsslcertificate_openssl.cpp:250:36: error: member access into incomplete
      type 'EVP_PKEY' (aka 'evp_pkey_st')
    } else if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_DH) {
                                   ^
/data/data/com.termux/files/usr/include/openssl/ossl_typ.h:93:16: note: 
      forward declaration of 'evp_pkey_st'
typedef struct evp_pkey_st EVP_PKEY;
               ^
8 errors generated.
make[2]: *** [Makefile:21907: .obj/qsslcertificate_openssl.o] Error 1
make[2]: Leaving directory '/data/data/com.termux/files/home/phantomjs/src/qt/qtbase/src/network'
```

My research concludes that it is attempting to include a dead, no longer used version of openssl (1.0.0?), which is seemingly not available in the Termux repos.

There is a proposed solution here:
https://stackoverflow.com/questions/35488166/error-compiling-qt5-with-openssl-linked
But as a commenter noted, it does not indeed help.
Unless of course I applied said fix incorrectly, by adding:
``LIBS += -L/data/data/com.termux/files/usr/include/openssl/ -lssl -lcrypto -l`` 
to ``src/qt/qtbase/src/network/ssl/ssl.pri``.

Do you fine people have any suggestions as to how to proceed?

--- a/x11-packages/phantomjs/build.sh
+++ b/x11-packages/phantomjs/build.sh
@@ -1,0 +1,27 @@
+TERMUX_PKG_HOMEPAGE=https://phantomjs.org/
+TERMUX_PKG_DESCRIPTION="A headless WebKit scriptable with JavaScript"
+TERMUX_PKG_LICENSE="BSD 3-Clause"
+TERMUX_PKG_LICENSE_FILE="LICENSE.BSD"
+TERMUX_PKG_MAINTAINER="@termux"
+_COMMIT=0a0b0facb16acfbabb7804822ecaf4f4b9dce3d2
+TERMUX_PKG_VERSION=2020.07.13
+TERMUX_PKG_SRCURL=https://github.com/ariya/phantomjs.git
+TERMUX_PKG_GIT_BRANCH=master
+TERMUX_PKG_DEPENDS="qt5-qtbase, qt5-qtwebkit"
+TERMUX_PKG_BUILD_DEPENDS="qt5-qtbase-cross-tools"
+TERMUX_PKG_FORCE_CMAKE=true
+
+# qt5-qtwebkit is not available for i686.
+TERMUX_PKG_BLACKLISTED_ARCHES="i686"
+
+termux_step_post_get_source() {
+	git fetch --unshallow
+	git checkout $_COMMIT
+
+	local version="$(git log -1 --format=%cs | sed 's/-/./g')"
+	if [ "$version" != "$TERMUX_PKG_VERSION" ]; then
+		echo -n "ERROR: The specified version \"$TERMUX_PKG_VERSION\""
+		echo " is different from what is expected to be: \"$version\""
+		return 1
+	fi
+}

